File: homework1/src/warmup.js
function change(cents) {
  if (cents < 0) {
    throw new RangeError();
  }

  const counts = [0, 0, 0, cents];
  const americanCoins = [25, 10, 5];

  for (let i = 0; i < 3; i += 1) {
    counts[i] = Math.floor(counts[3] / americanCoins[i]);
    counts[3] -= counts[i] * americanCoins[i];
  }

  return (counts);
}

function stripQuotes(quote) {
  return quote.replace(/["']/g, '');
}

function scramble(string) {
  let scrambled = '';
  const toScramble = string.split('');

  while (toScramble.length > 0) {
    scrambled += toScramble.splice(Math.round(toScramble.length * Math.random()), 1);
  }

  return scrambled;
}

function powers(base, limit, callback) {
  let power = 0;
  let outputPower = base ** power;
  power += 1;

  while (outputPower <= limit) {
    callback(outputPower);
    outputPower = base ** power;
    power += 1;
  }
}

function* powersGenerator(base, limit) {
  let power = 0;
  let outputPower = base ** power;
  power += 1;

  while (outputPower <= limit) {
    yield (outputPower);
    outputPower = base ** power;
    power += 1;
  }
}

function say(word) {
  if (word === undefined) {
    return '';
  }
  return function curry(otherWord) {
    if (otherWord === undefined) {
      return word;
    }
    return say(`${word} ${otherWord}`);
  };
}

function interleave(array, ...values) {
  const interleaved = [];

  for (let i = 0, j = 0; i < array.length || j < values.length;) {
    if (i < array.length) {
      interleaved.push(array[i]);
      i += 1;
    }
    if (j < values.length) {
      interleaved.push(values[j]);
      j += 1;
    }
  }

  return interleaved;
}

function cylinder(spec) {
  let { radius, height } = spec;

  if (radius === undefined) {
    radius = 1;
  }
  if (height === undefined) {
    height = 1;
  }

  const surfaceArea = () => 2 * Math.PI * radius * height + 2 * Math.PI * radius * radius;
  const volume = () => Math.PI * radius * radius * height;
  const widen = (factor) => { radius *= factor; };
  const stretch = (factor) => { height *= factor; };
  const toString = () => `Cylinder with radius ${radius} and height ${height}`;
  return Object.freeze({
    surfaceArea,
    volume,
    get radius() { return radius; },
    get height() { return height; },
    widen,
    stretch,
    toString,
  });
}

const crypto = require('crypto');

function makeCryptoFunctions(key, algorithm) {
  const funcs = [];
  funcs[0] = function encrypt(utf8) {
    const theKey = crypto.createCipher(algorithm, key);
    let hex = theKey.update(utf8, 'utf8', 'hex');
    hex += theKey.final('hex');
    return hex;
  };
  funcs[1] = function decrypt(hex) {
    const theKey = crypto.createDecipher(algorithm, key);
    let utf8 = theKey.update(hex, 'hex', 'utf8');
    utf8 += theKey.final('utf8');
    return utf8;
  };
  return funcs;
}

const request = require('request');

function randomName(info) {
  return new Promise((resolve, reject) => {
    request(`https://uinames.com/api/?amount=1&gender=${info.gender}&region=${info.region}`, (error, response, body) => {
      if (error || response.statusCode !== 200) reject(new Error(response.statusCode));
      else resolve(`${body.surname}, ${body.name}`);
    });
  });
}

module.exports = {
  change,
  stripQuotes,
  scramble,
  cylinder,
  randomName,
  say,
  makeCryptoFunctions,
  interleave,
  powers,
  powersGenerator,
};

File: homework2/warmup.py
import math
import random
from cryptography.fernet import Fernet
import base64
import requests

def change(cents):
    if cents < 0:
        raise ValueError('amount cannot be negative')
    count = ([0, 0, 0, cents])
    american_coins = ([25, 10, 5])
    for i in range(3):
        count[i] = math.floor(count[3] / american_coins[i])
        count[3] -= count[i] * american_coins[i]
    return tuple(count)


def strip_quotes(quote):
    return quote.replace('"', '').replace("'", '')


def scramble(string):
    return ''.join(random.sample(string, len(string)))


def say(word=None):
    if word == None:
        return ''

    def curry(otherWord=None):
        if otherWord == None:
            return curry.string
        curry.string += " " + otherWord
        return curry
    curry.string = word
    return curry


def triples(n):
    return [(c, b, a) for a in range(n+1) for b in range(a) for c in range(b) if a*a == b*b + c*c]


def powers(base, limit):
    power = 0
    outputPower = base ** power
    power += 1

    while outputPower <= limit:
        yield (outputPower)
        outputPower = base ** power
        power += 1


def interleave(a, *b):
    firstLength = len(a)
    secondLength = len(b)
    maximum = max(firstLength, secondLength)
    result = []
    for i in range(maximum):
        if i < firstLength:
            result.append(a[i])
        if (i < secondLength):
            result.append(b[i])
    return result

class Cylinder:
    def __init__(self, radius=1, height=1):
        self.radius = radius
        self.height = height
        self.volume = self.radius * self.radius * math.pi * self.height
        self.surface_area = 2 * math.pi * self.radius * \
            self.height + 2 * math.pi * self.radius * self.radius

    def widen(self, factor):
        self.radius *= factor
        self.volume = self.radius * self.radius * math.pi * self.height
        self.surface_area = 2 * math.pi * self.radius * \
            self.height + 2 * math.pi * self.radius * self.radius
        return self.radius

    def stretch(self, factor):
        self.height *= factor
        self.volume = self.radius * self.radius * math.pi * self.height
        self.surface_area = 2 * math.pi * self.radius * \
            self.height + 2 * math.pi * self.radius * self.radius
        return self.height


def make_crypto_functions(theKey):
    key = Fernet(theKey)
    def e(message):
        return key.encrypt(message)
    def d(message):
        return key.decrypt(message)
    return e, d

def random_name(**kwargs):
    params = {'amount': 1}
    if 'gender' in kwargs:
        params['gender'] = kwargs['gender']
    if 'region' in kwargs:
        params['region'] = kwargs['region']
    
    url = 'https://uinames.com/api'
    response = requests.get(url, params=params)

    if response.status_code not in range(200, 300):
        raise ValueError(response.text)

    data = response.json()

    return data['surname'] + ', ' + data['name']